qcengine/programs/madness/harvester.py

```diff
@@ -0,0 +1,222 @@
+import re
+import json
+import logging
+from decimal import Decimal
+from typing import Tuple
+
+import numpy as np
+import qcelemental as qcel
+from qcelemental.models import Molecule
+from qcelemental.models.results import AtomicResultProperties
+from qcelemental.molparse import regex
+
+from ..util import PreservingDict
+
+logger = logging.getLogger(__name__)
+
+
+def harvest_output(outtext: str) -> Tuple[PreservingDict, Molecule, list, str, str]:
+    """Function to read an entire MADNESS output file.
+
+    Reads all of the different "line search" segments of a file and returns
+    values from the last segment for which a geometry was written.
+
+    Args:
+        outtext (str): Output written to stdout
+    Returns:
+        - (PreservingDict) Variables extracted from the output file in the last complete step
+        - (Molecule): Molecule from the last complete step
+        - (list): Gradient from the last complete step
+        - (str): Version string
+        - (str): Error message, if any
+    """
+
+    # Loop over all steps
+    # TODO (wardlt): Is it only necessary to read the last two steps?
+    pass_psivar = []
+    pass_coord = []
+    pass_grad = []
+    for outpass in re.split(r"Converged!", outtext, re.MULTILINE):
+        psivar, madcoord, madgrad, version, error = harvest_outfile_pass(outpass)
+        pass_psivar.append(psivar)## all the variables extracted
+        pass_coord.append(madcoord)
+        pass_grad.append(madgrad)
+
+    # Determine which segment contained the last geometry
+    retindx = -1 #if pass_coord[-1] else -2
+    return pass_psivar[retindx], pass_coord[retindx], pass_grad[retindx], version, error
+
+
+def harvest_outfile_pass(outtext):
+    """Function to read Madness output file *outtext* and parse important
+    quantum chemical information from it in
+
+    """
+    psivar = PreservingDict()
+    psivar_coord = None
+    psivar_grad = None
+    version = ""
+    error = ""  # TODO (wardlt): The error string is never used.
+
+    NUMBER = r"(?x:" + regex.NUMBER + ")"
+    # fmt: off
+
+    # Process version
+    mobj = re.search(
+        r'^\s+' + r'MADNESS' + r'\s+' + r'(\d+.\d\d+.\d)' +r'\s'+ r'multiresolution suite'+r'\s*$',
+        outtext, re.MULTILINE)
+    if mobj:
+        logger.debug('matched version')
+        version = mobj.group(1)
+
+    # Process SCF
+    # 1)Fail to converge (TODO Robert ask for failed convergence)
+    mobj = re.search(r'^\s+' + r'(?:Calculation failed to converge)' + r'\s*$', outtext, re.MULTILINE)
+    if mobj:
+        logger.debug('failed to converge')
+
+    # 2)Calculation converged
+    else:
+        OPTIONS=[r'kinetic',r'nonlocal psp',r'nuclear attraction',r'coulomb',r'PCM',r'exchange-correlation',r'nuclear-repulsion',r'total']
+        PSIVAR=['KINETIC ENERGY','NONLOCAL PSP','NUCLEAR ATTRACTION ENERGY','COULOMB','PCM','EXCHANGE-CORRELATION','NUCLEAR REPULSION ENERGY','TOTAL SCF ENERGY']
```

Comment:
maybe collect into AtomicResultProperties directly, since those names will get validated. these aren't "official" psivars/qcvars (but the code that checks names isn't in qcengine).